File: slidingWindow.js
function maxProfit(prices) {
  let buy = 0;
  let max_profit = 0;
  // Start from the second day
  for (let sell = 1; sell < prices.length; sell++) {
    // Check if the current day's price greater than the buying day's price
    if (prices[sell] > prices[buy]) {
      // update profit
      let profit = prices[sell] - prices[buy];
      max_profit = Math.max(max_profit, profit);
    } else {
      // If the selling price  is lower than at buying, update the buying day
      buy = sell;
    }
  }
  return max_profit;
}
// console.log(maxProfit([10, 1, 5, 6, 7, 1]));

function lengthOfLongestSubstring(s) {
  let char_set = new Set();
  let left = 0;
  let length = 0;
  for (let right = 0; right < s.length; right++) {
    // If the character at 'right' is already in the set,
    // Remove characters until duplicates and move left
    while (char_set.has(s[right])) {
      char_set.delete(s[left]);
      left++;
    }
    char_set.add(s[right]);
    length = Math.max(length, right - left + 1);
  }
  return length;
}
// console.log(lengthOfLongestSubstring('abfcaczxaywkl'));

function characterReplacement(s, k) {
  let left = 0;
  let count = new Map();
  let result = 0;

  for (let right = 0; right < s.length; right++) {
    let len = right - left + 1;
    // Update the count of the current character
    count.set(s[right], 1 + (count.get(s[right]) || 0));

    // checks if the number of characters in the window that would need to be replaced exceeds k.
    // If it does, it means the window is no longer valid, so we need to shrink it.
    if (len - Math.max(...count.values()) > k) {
      // Reduce the count of the character at the left pointer (s[left])
      count.set(s[left], count.get(s[left]) - 1);
      left++;
    }
    // Update the new window length after adjustment
    len = right - left + 1;

    result = Math.max(result, len);
  }
  return result;
}
// console.log(characterReplacement('XYYZX', 2));
function checkInclusion(s1, s2) {
  if (s1.length > s2.length) return false;

  let neededChar = {}; //To Store the frequency/count of required string s1
  for (let i = 0; i < s1.length; i++) {
    let el = s1[i];
    neededChar[el] = (neededChar[el] || 0) + 1;
    console.log(neededChar);
  }

  let left = 0, //left pointer
    right = 0, //right pointer
    requiredLength = s1.length; //substring length required in s2
  console.log('ok');

  //  iterate until the right pointer of window is less than s2 length
  while (right < s2.length) {
    // If we found s2 character in s1 / neededChar then we decrease requiredLength
    if (neededChar[s2[right]] > 0) {
      requiredLength--;
    }
    // Since we have encountered new char i.e s2[right] we decrease it's
    // count in neededChar even if it is not present in neededChar because we only care about neededChars
    neededChar[s2[right]]--;
    console.log(neededChar);
    right++;
    // Now if our requiredLength becomes 0 it means we have found a match of the s2 substring
    // So we return true
    if (requiredLength === 0) return true;

    // If the window size equals s1's length, slide the window
    if (right - left === s1.length) {
      // if the left element we are removing was a required character then we increase requiredLength
      // because that element will no longer be the part of sliding window
      if (neededChar[s2[left]] >= 0) requiredLength++;
      // We will also increase the count of left element removed from window
      neededChar[s2[left]]++;
      // And finally we will decrease the window size by 1 from left i.e left++
      left++;
    }
  }
  return false;
}

console.log(checkInclusion('ab', 'eidbaooo'));

File: slidingWindow.py
import collections


def maxProfit(prices):
    buy_price = prices[0]
    max_profit = 0
    for price in prices:
        buy_price = min(buy_price, price)
        max_profit = max(max_profit, price - buy_price)
    return max_profit


# prices = [10, 8, 7, 5, 2]
# print(maxProfit(prices))


def minWindow(s: str, t: str) -> str:
    if not s or not t or len(s) < len(t):
        return ""
    t_char_count = {}

    for c in t:
        t_char_count[c] = 1 + t_char_count.get(c, 0)

    window = {}
    required = len(t_char_count)
    minLen = float("inf")
    left = 0
    start, end = 0, 0

    for right in range(len(s)):
        char = s[right]
        window[char] = 1 + window.get(char, 0)
        if char in t_char_count and window[char] == t_char_count[char]:
            required -= 1
        # if required is 0, the current window is valid
        while required == 0:
            if right - left + 1 < minLen:
                start = left
                end = right + 1
                minLen = right - left + 1
            window[s[left]] -= 1

            if s[left] in t_char_count and window[s[left]] < t_char_count[s[left]]:
                required += 1
            left += 1

    return s[start:end]


# print(minWindow("abcdxyreoy", "xyoy"))


def maxSlidingWindow(nums, k):
    output = []
    q = collections.deque()
    l = r = 0
    while r < len(nums):

        while q and nums[q[-1]] < nums[r]:
            print(nums[q[-1]], nums[r])
            q.pop()
        q.append(r)

        if l > q[0]:
            q.popleft()

        if (r + 1) >= k:
            output.append(nums[q[0]])
            l += 1
        r += 1
    print(output)
    return output

    # maxSlidingWindow([1, 2, 1, 0, 4, 2, 6], 3)


def checkInclusion(s1: str, s2: str) -> bool:
    if len(s1) > len(s2):
        return False
    # Count characters in s1
    s1_count = {}
    for c in s1:
        s1_count[c] = s1_count.get(c, 0) + 1
    # Track how many unique characters we still need to match
    required_matches = len(s1_count)
    l = 0
    print(s1_count)
    print(required_matches)
    for r in range(len(s2)):
        # Add character from right

        print(s2[r])

        if s2[r] in s1_count:
            s1_count[s2[r]] -= 1
            if s1_count[s2[r]] == 0:
                required_matches -= 1
        # Check if we have a valid window
        print(s1_count)
        print(required_matches)
        if required_matches == 0:
            return True
        # If window size equals s1 length, move left pointer
        print("Left=", l, "Right=", r)

        if r - l + 1 == len(s1):
            print("Window")
            if s2[l] in s1_count:
                if s1_count[s2[l]] == 0:

                    required_matches += 1
                s1_count[s2[l]] += 1
                print(s1_count)
                print(required_matches)
            l += 1
    return False


print(checkInclusion("hello", "ooolleoooleh"))
print(checkInclusion("abc", "lescabb"))
